### PR TITLE
[Arduino Core 3.0.0] RMT IDF5.1 Refactoring

### PR DESCRIPTION
**Pull request by SuGlider**
## Description of Change
Arduino RMT Refactoring.

Documentation also must be updated.

## Tests scenarios
Tested on ESP32, ESP32S2, ESP32C3 and ESP32S3 using all the examples here provided.

### CODE REVIEW

**Comment by PilnyTomas** [comment]
Compiling example Blink and BlinkRGB produces error: `esp32/cores/esp32/esp32-hal-cpu.c:244:108: error: 'RTC_CPU_FREQ_SRC_APLL' undeclared` But I can see that this file was not changed...

**Comment by SuGlider** [comment]
> Compiling example Blink and BlinkRGB produces error: `esp32/cores/esp32/esp32-hal-cpu.c:244:108: error: 'RTC_CPU_FREQ_SRC_APLL' undeclared` But I can see that this file was not changed...

I have tested the example from https://github.com/espressif/arduino-esp32/blob/master/libraries/ESP32/examples/GPIO/BlinkRGB/BlinkRGB.ino

It works as expected.

@PilnyTomas - Please make sure you checkout the correct branch in order to compile and test it.
https://github.com/espressif/arduino-esp32/tree/esp-idf-v5.1-libs

**Comment by SuGlider** [comment]
@me-no-dev -- PTAL

File: cores/esp32/esp32-hal-rmt.c [hunk 1]
```diff
@@ -13,622 +13,556 @@
...
+
+  // validate the RMT ticks by the requested frequency
+  // Based on 80Mhz using a divider of 8 bits (calculated as 1..256)
+  if (frequency_Hz > 80000000 || frequency_Hz < 312500) {
+    log_e("GPIO %d - Bad RMT frequency resolution. Must be between 312.5KHz to 80MHz.", pin);
+    return false;
+  }
+
+  // Try to dettach any (Tx|Rx|Whatever) previous bus or just keep it as not attached
+  if (perimanGetPinBusType(pin) != ESP32_BUS_TYPE_INIT && !perimanSetPinBus(pin, ESP32_BUS_TYPE_INIT, NULL)) {
```
Comment (P-R-O-C-H-Y) [inline cores/esp32/esp32-hal-rmt.c:447]: What if `perimanGetPinBusType(pin)` will return `ESP32_BUS_TYPE_INIT`, then the IF will fail. I think that part of checking `perimanGetPinBusType(pin) != ESP32_BUS_TYPE_INIT` can be removed and just leave the second part of IF.

File: cores/esp32/esp32-hal-rmt.h [hunk 2]
```diff
@@ -19,151 +19,199 @@
...
 /**
- * Setting input filter
- */
-bool rmtSetFilter(rmt_obj_t* rmt, bool filter_en, uint32_t filter_level);
+   Function used to filter input noise in the RX channel.
+   In receiving mode, channel will ignore any input pulse which width is smaller than <filter_pulse_ns>
+   <filter_en> is used to enable/disable the filter. <true> enables it, <false> disables.
+   The function returns <true> if it is correctly executed, <false> otherwise.
+*/
+bool rmtSetFilter(int pin, bool filter_en, uint8_t filter_level);
```
[the remark this answers is not in the document]
Comment (SuGlider) [reply cores/esp32/esp32-hal-rmt.h:205]: ok. I'll change it.

**Pull request merged**